### ISSUE TITLE
Version 3.0beta is broken in many ways

### ISSUE THREAD

**Issue by MaxSavenkov** [eddic/fastcgipp#9]
1) The code to set non-blocking mode sets blocking mode instead. Re-read the one-liner you use:

fcntl(listen, F_SETFL, (fcntl(listen, F_GETFL)|O_NONBLOCK)^O_NONBLOCK);

It gets current socket flags, appends O_NONBLOCK to them with logic OR, and THEN, for some reason, does XOR with O_NONBLOCK, which leads to dropping of O_NONBLOCK flag).

2) Several places where invalid iterator (end) is dereferenced:
- Fastcgipp::FcgiStreambuf<wchar_t, std::char_traits<wchar_t>>::emptyBuffer()
  &*record.end() - it's wrong to do this.
- void Fastcgipp::Http::vecToString(
  string = converter.from_bytes(&_start, &_end); - DITTO

3) Bad reinterpret cast:
_reinterpret_cast<uint16_t_>(write) = 0xffff; 
should be
_reinterpret_cast<uint16_t_>(&*write) = 0xffff;

4) Vector iterator reused between different vectors (which is not allowed):

In template<class charT> void Fastcgipp::Http::decodeUrlEncoded(

nameEnd is first used as 

auto nameEnd(dataEnd);

but later is resued in this statement:

```
            nameEnd = percentEscapedToRealBytes(
                    nameStart,
                    data,
                    buffer.begin());
```

Where it becomes an iterator from "buffer".

I'm sure there are many more errors here, but at this point, I kind gave up, this library is not ready to be used in production.

**Comment by eddic** [comment]
Good catches. Thanks. I guess I ended up putting some ugly non-standard stuff in the codebase without catching it. I'm surprised GCC never complained about all that stuff with _-Wall -Werror_.

I'm amazed that NONBLOCK bug has lived so long so broken. The reinterpret_cast is weird. I don't understand why the compiler didn't complain about that. It's almost as though GCC allows you to reinterpret_cast a vector interator as though it's a pointer. I guess in this implementation vector iterators _are_ just pointers and GCC is allowing all that weird non-standard stuff to pass as a result.

As far as the last one you mention. I fail to see what is wrong with re-assigning a vector<char>::iterator variable across multiple containers. I've yet to find anything specific that would indicate this as having undefined behaviour.

> I'm sure there are many more errors here, but at this point, I kind gave up, this library is not ready to be used in production.

Come on man. It is marked as beta for a reason. I'm not sure what would have led you to believe it is ready for production. I'll make sure to send you a full refund for the $0.00 you paid for the library. I'll never understand the over-inflated senses of entitlement people have for some FOSS code they found online. I appreciate your feedback on the specific issues you've found but I could not care less that Max Savenkov thinks the library is not "ready for production". Feel free to move on to another project and take your unwarranted expectations with you.

**Comment by Erroneous1** [comment]
I've found that -Wall doesn't show nearly enough warnings for C++.

Personally I like the following:
`-Werror -Wold-style-cast -Wnon-virtual-dtor -Wshadow -Winit-self -Wredundant-decls -Wfloat-equal -Wunreachable-code -Wmissing-declarations -Wmissing-include-dirs -Wswitch-enum -Weffc++ -Wzero-as-null-pointer-constant -Wsuggest-override -Wsuggest-attribute=const -Wsuggest-attribute=pure -Wduplicated-cond -Wtrampolines -Wconversion -Wunsafe-loop-optimizations -Wplacement-new -Wpointer-arith -Wtype-limits -Wcast-align -Wdouble-promotion -Wuninitialized -Wparentheses -Wuseless-cast -Wlogical-op -Wmissing-field-initializers -Winline -pedantic -Wextra -Wall -fexceptions`

Some things like the -Wsuggest-* options are a matter of taste, but it does make code more explicit. -Wuninitialized, -Wzero-as-null-pointer-constant, -Wconversion, and -Wshadow remind me of some of the errors you get with Java. BTW -Weffc++ always gave me errors on the 2.1 release because of non-virtual destructors. At minimum I like -Wall -Wpedantic -Weffc++.

**Comment by MaxSavenkov** [comment]
Sorry, I just was pretty strongly frustrated after spending two workdays porting the code to work on Windows, and then getting all these non-platform-related errors. The problem is, I need library that is production-ready, and the choice is very, very limited - few people do FastCGI, and fewer still use C/C++. I moved on to the "official" libfcgi for now, and it seems to work, but its API a bit too Spartan in for my tastes.

One thing Microsoft compiler (or, rather, its standard library) is good at, is catching errors in iterator usage in STL, due to presence of iterator debugging feature - this was how I found some of these problems. It also treats assignment of iterators between containers as a runtime error. There is a thread on SO (http://stackoverflow.com/questions/4657513/comparing-iterators-from-different-containers) which discusses this, and comes to conclusion that such comparison is undefined.

reinterpret_case one, I think, is highly implementation-dependent.

Have you tried running static analysis tools on your code? It often does wonders where compilers fail. I used cppcheck, and even trial version of PVS-Studio on some projects, and they catch errors that would otherwise take quite a lot of time to notice, and if you add Clang analysis to that, you can probably cover most statically-catchable bugs.

**Comment by aggsol** [comment]
The Version number 3.0 suggests a certain maturity of the library
@MaxSavenkov What lib exactly are you using now? I am looking for one, too. No it seems I am stuck with cgicc (https://www.gnu.org/software/cgicc/)

**Comment by MaxSavenkov** [comment]
Well, it's 3.0beta, so yeah, I probably should not have expected too much of it... But I was seduced by use of CMake instead of autotools and lack of Boost dependency compated to 2.1. I mean, a lot of OSS never gets past 1.0, and yet everyone is using it.

@aggsol I'm using https://github.com/eatonphil/fastcgi-development-kit (this repo is unofficial, but official repo is dead). Cgicc looks like it has more features, but I don't need most of those (e.g. HTML constructor).

**Comment by Erroneous1** [comment]
You can use libfcgi and mix it with the LGPL Gnu cgicc library. There is even an example that shows how to use it with fastcgi. I agree that doing the html stuff is unnecessary (kind of like the Perl CGI module). Personally I think it should be a separate library and that proper web design should probably use. Both of these libraries are mature.

I would also consider helping out this project because I agree that a good modern fastcgi library would be nice. I don't like that fastcgi.com is gone even if it was stable.

**Comment by eddic** [comment]
Yeah I suppose I overreacted. Apologies for that. I guess I should make it a little more obvious that 3.0 has still not been adequately tested for reliable deployment. I myself have no immediate plans to migrate my critical web applications to 3.0 until I'm confident in it's stability. All the paid web services I offer still use 2.1. I have 3.0 running for some fun stuff for testing purposes but nothing critical. As far as 3.0 implying maturity, keep in mind that 3.0 is effectively a total rewrite of the library. I really do think that 3.0 has some great potential and is already pretty stable. I feel that it's ultimately one of the better solutions for C++/FastCGI out there but I am, of course, very biased.

Also keep in mind that I have no motivation to get the code running on Windows. My use of the library is strictly on GNU/Linux environments compiled with GCC. Any attempts I made for portability were not in anyway tested beyond said environment. I'm not going to use my time getting fastcgi++ working on a platform I'll never use. That being said I'm happy to accommodate those that wish to do so as it'll only result in a more portable codebase that will age better.

So what are my motivations? Perhaps I'm not upfront enough with them. I would really like to start migrating my critical web applications to 3.0 come summertime so I'm hoping the community can assist in debugging the library. Really I'm just here to trade code for feedback for my own selfish reasons. I mean, isn't that really what this is about for most of us? I'm currently embroiled in other projects but I'm hoping to have accumulated enough feedback and data to make some large improvements come spring/summer.

So upon further examination of that whole iterator reassignment, I see where my error was. The issue was really not in reassignment. The issue was my using said iterator to store parser state information and later doing a comparison across containers. I dig that being a problem and will adjust the design.

In conclusion. I really do appreciate all of you and the feedback you bring. Don't give up and fastcgi++ quite yet!
